packages/abi-utils/lib/normalize.ts

```diff
@@ -0,0 +1,47 @@
+import { Abi as SchemaAbi } from "@truffle/contract-schema/spec";
+import { Abi, Entry, StateMutability } from "./types";
+
+//NOTE: SchemaAbi is kind of loose and a pain to use.
+//So we'll generally coerce things to Abi before use.
+//(we combine this with adding a "function" tag for
+//entries that lack it)
+
+export const normalize = (looseAbi: SchemaAbi | Abi): Abi =>
+  (looseAbi as SchemaAbi).map(normalizeEntry);
+
+type Item<A> = A extends (infer I)[] ? I : never;
```

Comment:
I'm worried about adding this to that package because those types are autogenerated and I don't want to break them accidentally

Reply:
To be clear, what I mean by "adding that type" is factoring that type out in the schema, giving it its own definition there, not manually editing the Typescript or something.

Reply:
Just tried it - it doesn't work. `json2ts` seems to only respect things that are defined at the top-level; it doesn't emit separate types for individual definitions.

Reply:
Yes, by "add that type" what I'm saying is you should factor out that type as a top-level definition, like we have for so many other types!

Reply:
That's what I did - it doesn't work. Notice how you can't do `import { NormalFunction } from "@truffle/contract-schema/spec"`. json2ts just doesn't create dedicated types for individual definitions.

Reply:
I'm confused.  You can import `ImmutableReferences`, as an example.  What's different about this?

Reply:
no idea, but I tried it and it doesn't show up.

Reply:
can we call that out of scope for this PR?

Reply:
Eh, I guess?  I think it would be good to do (it's something I'd been thinking about doing myself for some time, because it's always been an inconvenience) but if it doesn't work for some reason then I guess it's not essential.

Reply:
https://github.com/trufflesuite/truffle/issues/3431